suite/tests/client-interface/drx-scattergather.dll.c

```diff
@@ -139,14 +140,14 @@ event_bb_app2app(void *drcontext, void *tag, instrlist_t *bb, bool for_trace,
             scatter_gather_present = true;
         } else if (instr_is_scatter(instr)) {
             scatter_gather_present = true;
-        } else if (!one_time_mask_clobber_test) {
-            if (instr_is_mov_constant(instr, &val) && val == TEST_MASK_CLOBBER_MARKER) {
-                instr_t *next_instr = instr_get_next(instr);
-                if (next_instr != NULL) {
-                    if (instr_is_mov_constant(next_instr, &val) &&
-                        val == TEST_MASK_CLOBBER_MARKER) {
-                        one_time_mask_clobber_test = true;
-                    }
+        } else if (instr_is_mov_constant(instr, &val) &&
+                   val == TEST_MASK_CLOBBER_MARKER) {
+            instr_t *next_instr = instr_get_next(instr);
+            if (next_instr != NULL) {
+                if (instr_is_mov_constant(next_instr, &val) &&
+                    val == TEST_MASK_CLOBBER_MARKER) {
+                    one_time_mask_clobber_test = true;
+                    one_time_mask_clobber_translating = true;
```

Comment:
Shouldn't this only set one or the other, based on `translating`?

Reply:
No, because the fault happens on the gather instruction, which is after the marker, so the translation event doesn't see the markers. I think this is because gather instructions are broken into separate bbs for expansion? The fault happens right after the translation was installed and executes. This should be safe.

Reply:
I don't understand how it works: a single condition here sets both _test and _translating.  Later, only one of these is cleared.  So one is left true.  Having no condition here when setting both, yet having a condition when clearing, does not look right.

Reply:
Originally, I only had one flag for the test, which is enough for it to work. I then changed it to 2 flags upon your request. The reason that I can't set the translation flag to true only for "translating" is that it the markers never appear in the list when the translating, because the fault happens in the block of the gather instructions. So not sure how to change this in a way that you're happy.

Side note: I tried to write this test with signal but it seemed futile to try hitting the small corner case window.

Reply:
I never requested any flags or flags arrangement: I am requesting clear and robust code, or at the least to have comments explaining why it's fragile but we're ok living with it.  It seems to satisfy neither of those conditions.  Originally there was a global flag for identifying a particular code location, but it was never reset, so any kandnw after that point will be treated as the target.  Now there is still a global that is never reset (the _translating flag).  It is not any better: it seems a step backward, since *both* globals are set, but then only *one* is reset, so it has similar problems as before but with more complexity and confusion for the reader.

If someone takes this code as a base for some more general non-test code (most code starts out modeled on some other existing code), it's going to have problems while the reader puzzles over how this code works, with this unusual asymmetric flag arrangement.  If there is no simple way to communicate between the identification and use, we should consider adding or tweaking the interface to make that possible.  And, there should be *comments* explaining why this is not ideal code and that it should not be used as a model for applying to anything but this particular test app.  This is a general principle: always clearly label unfinished or flawed code!  Do not shoot future code maintainers in the foot.

Could the app itself already have a ud2 there?  It says you're using the *app's* signal handler.

For the communication problem with the split bb: adding a label instruction or using the propagated user_data don't work b/c of the brand-new bb.  Maybe record the PC of the gather?  When the pattern is seen, decode forward and if a gather is found before any cti, save its PC.  That seems much better than a boolean flag.

Reply:
> So not sure how to change this in a way that you're happy.

Are *you* happy with this flags code??

Reply:
> Are you happy with this flags code??

I was, since it was one flag on/off for bb creation, and another for the translating event (Re: but it was never reset: ? both were set to false).

I've changed using flags to doing decode and using the gather pc to identify the injection point.

Done.

Reply:
Re: Could the app itself already have a ud2 there? It says you're using the app's signal handler.

No, the ud2 is in the *expanded* sequence. You understand the test right? We need to get a recreate state event within a tiny window in the long expanded gather sequence somehow, and I am falling back to injecting a ud2 into the expanded sequence, because I tried using asynchronous signals and it seems hard or impossible to hit the small corner case window the test needs. As pointed out above, I have changed the flags into doing a decode + saving the gather pc in order to identify the first kandnw instruction as suggested. The behavior is exactly the same as before. Resolving, please re-open if necessary.

Reply:
Both were not reliably reset: it seemed to rely very heavily on particular app behavior, yet it had not a single comment stating its assumptions.  It looked fragile and not a clean approach, yet it was presented as solid, general code.

Reply:
It still relies on the app behavior of the test *and* the expansion part as well. I will add yet another comment mentioning this. I think this is a good idea.